File: 2022/day11/zig/solution.zig
const std = @import("std");

const MAX_MONKEYS = 10;
const MAX_ITEMS = 64;

const Operation = enum {
    add,
    mul,
};

const Monkey = struct {
    items: [MAX_ITEMS]u64,
    item_count: usize,
    op: Operation,
    operand: ?u64, // null means "old"
    divisor: u64,
    if_true: usize,
    if_false: usize,
    inspections: u64,

    fn init() Monkey {
        return Monkey{
            .items = undefined,
            .item_count = 0,
            .op = .add,
            .operand = 0,
            .divisor = 1,
            .if_true = 0,
            .if_false = 0,
            .inspections = 0,
        };
    }

    fn addItem(self: *Monkey, item: u64) void {
        self.items[self.item_count] = item;
        self.item_count += 1;
    }

    fn clearItems(self: *Monkey) void {
        self.item_count = 0;
    }

    fn applyOp(self: *const Monkey, old: u64) u64 {
        const val = self.operand orelse old;
        return switch (self.op) {
            .add => old + val,
            .mul => old * val,
        };
    }
};

fn parseNumber(line: []const u8, start: usize) struct { val: u64, end: usize } {
    var i = start;
    // Skip non-digits
    while (i < line.len and (line[i] < '0' or line[i] > '9')) : (i += 1) {}
    if (i >= line.len) return .{ .val = 0, .end = i };

    var val: u64 = 0;
    while (i < line.len and line[i] >= '0' and line[i] <= '9') : (i += 1) {
        val = val * 10 + (line[i] - '0');
    }
    return .{ .val = val, .end = i };
}

fn parseMonkeys(input: []const u8, monkeys: *[MAX_MONKEYS]Monkey) usize {
    var count: usize = 0;
    var i: usize = 0;

    while (i < input.len) {
        // Skip to "Monkey X:"
        while (i < input.len and !std.mem.startsWith(u8, input[i..], "Monkey ")) : (i += 1) {}
        if (i >= input.len) break;

        var monkey = Monkey.init();

        // Skip to next line (Starting items)
        while (i < input.len and input[i] != '\n') : (i += 1) {}
        i += 1;

        // Parse starting items
        while (i < input.len and input[i] != '\n') {
            const res = parseNumber(input, i);
            if (res.end > i) {
                monkey.addItem(res.val);
                i = res.end;
            } else {
                i += 1;
            }
        }
        i += 1;

        // Parse operation line
        while (i < input.len and input[i] != '\n') {
            if (std.mem.startsWith(u8, input[i..], "old * old")) {
                monkey.op = .mul;
                monkey.operand = null;
                break;
            } else if (std.mem.startsWith(u8, input[i..], "old + ")) {
                monkey.op = .add;
                const res = parseNumber(input, i + 6);
                monkey.operand = res.val;
                break;
            } else if (std.mem.startsWith(u8, input[i..], "old * ")) {
                monkey.op = .mul;
                const res = parseNumber(input, i + 6);
                monkey.operand = res.val;
                break;
            }
            i += 1;
        }
        while (i < input.len and input[i] != '\n') : (i += 1) {}
        i += 1;

        // Parse divisor
        const div_res = parseNumber(input, i);
        monkey.divisor = div_res.val;
        i = div_res.end;
        while (i < input.len and input[i] != '\n') : (i += 1) {}
        i += 1;

        // Parse if_true target
        const true_res = parseNumber(input, i);
        monkey.if_true = @intCast(true_res.val);
        i = true_res.end;
        while (i < input.len and input[i] != '\n') : (i += 1) {}
        i += 1;

        // Parse if_false target
        const false_res = parseNumber(input, i);
        monkey.if_false = @intCast(false_res.val);
        i = false_res.end;

        monkeys[count] = monkey;
        count += 1;
    }

    return count;
}

fn simulate(monkeys: *[MAX_MONKEYS]Monkey, count: usize, rounds: usize, relief_divisor: u64, mod_value: u64) void {
    for (0..rounds) |_| {
        for (0..count) |m| {
            const monkey = &monkeys[m];
            for (0..monkey.item_count) |item_idx| {
                var worry = monkey.items[item_idx];
                monkey.inspections += 1;

                // Apply operation
                worry = monkey.applyOp(worry);

                // Apply relief
                if (relief_divisor > 1) {
                    worry = worry / relief_divisor;
                }

                // Apply modulo to prevent overflow
                if (mod_value > 0) {
                    worry = worry % mod_value;
                }

                // Test and throw
                const target = if (worry % monkey.divisor == 0) monkey.if_true else monkey.if_false;
                monkeys[target].addItem(worry);
            }
            monkey.clearItems();
        }
    }
}

fn monkeyBusiness(monkeys: *const [MAX_MONKEYS]Monkey, count: usize) u64 {
    var inspections: [MAX_MONKEYS]u64 = undefined;
    for (0..count) |i| {
        inspections[i] = monkeys[i].inspections;
    }

    // Sort descending (simple bubble sort for small array)
    for (0..count) |i| {
        for (i + 1..count) |j| {
            if (inspections[j] > inspections[i]) {
                const tmp = inspections[i];
                inspections[i] = inspections[j];
                inspections[j] = tmp;
            }
        }
    }

    return inspections[0] * inspections[1];
}

fn copyMonkeys(src: *const [MAX_MONKEYS]Monkey, dst: *[MAX_MONKEYS]Monkey, count: usize) void {
    for (0..count) |i| {
        dst[i] = src[i];
    }
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const file_size = (try file.stat()).size;
    const input = try allocator.alloc(u8, file_size);
    defer allocator.free(input);

    _ = try file.readAll(input);

    // Parse monkeys
    var base_monkeys: [MAX_MONKEYS]Monkey = undefined;
    const count = parseMonkeys(input, &base_monkeys);

    // Calculate product of all divisors for part 2
    var mod_value: u64 = 1;
    for (0..count) |i| {
        mod_value *= base_monkeys[i].divisor;
    }

    // Part 1: 20 rounds with relief (divide by 3)
    var monkeys1: [MAX_MONKEYS]Monkey = undefined;
    copyMonkeys(&base_monkeys, &monkeys1, count);
    simulate(&monkeys1, count, 20, 3, 0);
    const part1 = monkeyBusiness(&monkeys1, count);

    // Part 2: 10000 rounds without relief
    var monkeys2: [MAX_MONKEYS]Monkey = undefined;
    copyMonkeys(&base_monkeys, &monkeys2, count);
    simulate(&monkeys2, count, 10000, 1, mod_value);
    const part2 = monkeyBusiness(&monkeys2, count);

    std.debug.print("Part 1: {d}\n", .{part1});
    std.debug.print("Part 2: {d}\n", .{part2});
}

File: 2022/day11/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#define MAX_MONKEYS 10
#define MAX_ITEMS 100

typedef struct {
    uint64_t items[MAX_ITEMS];
    int item_count;
    char op;           // '+' or '*'
    int operand;       // -1 means "old"
    int divisor;
    int if_true;
    int if_false;
    uint64_t inspections;
} Monkey;

int parse_monkeys(FILE *f, Monkey *monkeys) {
    char line[256];
    int monkey_count = 0;

    while (fgets(line, sizeof(line), f)) {
        if (strncmp(line, "Monkey", 6) == 0) {
            Monkey *m = &monkeys[monkey_count];
            m->item_count = 0;
            m->inspections = 0;

            // Starting items
            fgets(line, sizeof(line), f);
            char *p = strchr(line, ':') + 1;
            while (*p) {
                while (*p && (*p == ' ' || *p == ',')) p++;
                if (*p >= '0' && *p <= '9') {
                    m->items[m->item_count++] = (uint64_t)atoi(p);
                    while (*p >= '0' && *p <= '9') p++;
                } else {
                    break;
                }
            }

            // Operation
            fgets(line, sizeof(line), f);
            p = strstr(line, "old ");
            if (p) {
                p += 4;
                m->op = *p;
                p += 2;
                if (strncmp(p, "old", 3) == 0) {
                    m->operand = -1;
                } else {
                    m->operand = atoi(p);
                }
            }

            // Divisor
            fgets(line, sizeof(line), f);
            p = strstr(line, "by ");
            m->divisor = atoi(p + 3);

            // If true
            fgets(line, sizeof(line), f);
            p = strstr(line, "monkey ");
            m->if_true = atoi(p + 7);

            // If false
            fgets(line, sizeof(line), f);
            p = strstr(line, "monkey ");
            m->if_false = atoi(p + 7);

            monkey_count++;
        }
    }
    return monkey_count;
}

uint64_t apply_operation(uint64_t old, char op, int operand) {
    uint64_t val = (operand == -1) ? old : (uint64_t)operand;
    if (op == '+') {
        return old + val;
    } else {
        return old * val;
    }
}

void simulate(Monkey *monkeys, int count, int rounds, int relief_divisor, int use_modulo) {
    // Calculate product of all divisors for modulo
    uint64_t mod_value = 1;
    if (use_modulo) {
        for (int i = 0; i < count; i++) {
            mod_value *= monkeys[i].divisor;
        }
    }

    for (int round = 0; round < rounds; round++) {
        for (int i = 0; i < count; i++) {
            Monkey *m = &monkeys[i];
            while (m->item_count > 0) {
                uint64_t item = m->items[0];
                // Remove first item by shifting
                for (int j = 0; j < m->item_count - 1; j++) {
                    m->items[j] = m->items[j + 1];
                }
                m->item_count--;
                m->inspections++;

                // Apply operation
                uint64_t new_val = apply_operation(item, m->op, m->operand);

                // Apply relief
                if (relief_divisor > 1) {
                    new_val /= relief_divisor;
                }

                // Apply modulo
                if (use_modulo) {
                    new_val %= mod_value;
                }

                // Test and throw
                int target;
                if (new_val % m->divisor == 0) {
                    target = m->if_true;
                } else {
                    target = m->if_false;
                }
                monkeys[target].items[monkeys[target].item_count++] = new_val;
            }
        }
    }
}

uint64_t monkey_business(Monkey *monkeys, int count) {
    // Find top 2 inspection counts
    uint64_t top1 = 0, top2 = 0;
    for (int i = 0; i < count; i++) {
        if (monkeys[i].inspections > top1) {
            top2 = top1;
            top1 = monkeys[i].inspections;
        } else if (monkeys[i].inspections > top2) {
            top2 = monkeys[i].inspections;
        }
    }
    return top1 * top2;
}

void reset_monkeys(Monkey *dest, Monkey *src, int count) {
    for (int i = 0; i < count; i++) {
        dest[i] = src[i];
        dest[i].inspections = 0;
    }
}

int main() {
    FILE *f = fopen("../input.txt", "r");
    if (!f) {
        perror("Cannot open input.txt");
        return 1;
    }

    Monkey original[MAX_MONKEYS];
    int count = parse_monkeys(f, original);
    fclose(f);

    // Part 1
    Monkey monkeys[MAX_MONKEYS];
    reset_monkeys(monkeys, original, count);
    simulate(monkeys, count, 20, 3, 0);
    printf("Part 1: %llu\n", monkey_business(monkeys, count));

    // Part 2
    reset_monkeys(monkeys, original, count);
    simulate(monkeys, count, 10000, 1, 1);
    printf("Part 2: %llu\n", monkey_business(monkeys, count));

    return 0;
}
